spec.emu

```diff
@@ -1539,6 +1539,75 @@ return lastURL;</code></pre>
+  <h1>Operations on source map records</h1>
+
+  <p>After decoding a source map, source map consumers can use the resulting Decoded Source Map Records to look up position information for debugging or other use cases. This section describes the behaviour of typical operations that may be supported by source map consumers.</p>
+  <p>The GetOriginalPosition operation may be used to query the position in an original source that corresponds to a position in the generated code. For example, this can be used in a debugger to navigate from the generated code to the original source based on a user's mouse click.</p>
+  <p>The GetGeneratedPositionsForOriginalLine is used to query the positions in the generated code that map to a position in the original source. There may be multiple generated code positions that map to a given original position. The result list can be further filtered, for example, to match a particular column. This operation could be used, for example, as part of setting breakpoints at all generated lines that match a particular line in the original source that a user clicks in a debugger.</p>
+
+  <emu-clause id="sec-GetOriginalPosition" type="abstract operation">
+    <h1>
+      GetOriginalPosition (
+        _sourceMapRecord_: a Decoded Source Map Record,
+        _generatedPosition_: a Position Record,
+      ): a Decoded Mapping Record or *null*
+    </h1>
+    <dl class="header"></dl>
+    <emu-alg>
+      1. Let _mappings_ be _sourceMapRecord_.[[Mappings]].
+      1. Let _closest_ be *null*.
+      1. For each element _mapping_ of _mappings_, do
+        1. If the result of performing ComparePositions(_mapping_.[[GeneratedPosition]], _generatedPosition_) is ~greater~, then
+          1. Return _closest_.
+        1. Else,
+          1. Set _closest_ to _mapping_.
+      1. Return _closest_.
+    </emu-alg>
+  </emu-clause>
+
+  <emu-clause id="sec-GetGeneratedPositionsForOriginalLine" type="abstract operation">
+    <h1>
+      GetGeneratedPositionsForOriginalLine (
+        _sourceMapRecord_: a Decoded Source Map Record,
+        _originalSource_: a Decoded Source Record,
+        _originalLine_: a non-negative integral Number,
+      ): a list of Position Records
+    </h1>
+    <dl class="header"></dl>
+    <emu-alg>
+      1. Let _mappings_ be _sourceMapRecord_.[[Mappings]].
+      1. Let _generatedPositions_ be a new empty List.
+      1. Let _reverseMappings_ be ComputeReverseMappings(_mappings_, _originalSource_).
+      1. For each element _mapping_ of _reverseMappings_, do
+        1. Let _originalPosition_ be _mapping_.[[OriginalPosition]].
+        1. If _originalPosition_ ≠ *null*, then
+          1. If _originalPosition_.[[Line]] is equal to _originalLine_, then
+            1. Append _mapping_.[[GeneratedPosition]] to _generatedPositions_.
+      1. Return _generatedPositions_.
+    </emu-alg>
+
+    <emu-clause id="sec-ComputeReverseMappings" type="abstract operation">
+      <h1>
+        ComputeReverseMappings (
```

Comment:
Does this need to be a separate AO? We could just filter by `Decoded Mapping Record` and `originalLine` all at the same time.

Reply:
I don't think it has to be currently, the motivation for factoring out originally was that I intended to call the helper from other operations as well (if the spec has an operation that returns a single generated mapping for example).